File: h_lu.cpp
#include "any.h"
#include "low_rank.h"
#include "hierarchical.h"
#include "functions.h"
#include "batch.h"
#include "print.h"
#include "timer.h"

#include <algorithm>
#include <cmath>
#include <iostream>

using namespace hicma;

int main(int argc, char** argv) {
  int N = 1 << atoi(argv[2]);
  int nleaf = atoi(argv[3]);
  int rank = atoi(argv[4]);
  std::vector<double> randx(N);
  for (int i=0; i<N; i++) {
    randx[i] = drand48();
  }
  std::sort(randx.begin(), randx.end());
  start("Init matrix");
  int nblocks=0, admis=0;
  if (atoi(argv[1]) == 0) {
    nblocks = N / nleaf; // 1 level
    admis = N / nleaf; // Full rank
  }
  else if (atoi(argv[1]) == 1) {
    nblocks = 2; // Hierarchical (log_2(N/nleaf) levels)
    admis = N / nleaf; // Full rank
  }
  else if (atoi(argv[1]) == 2) {
    nblocks = 4; // Hierarchical (log_4(N/nleaf) levels)
    admis = N / nleaf; // Full rank
  }
  else if (atoi(argv[1]) == 3) {
    nblocks = N / nleaf; // 1 level
    admis = 0; // Weak admissibility
  }
  else if (atoi(argv[1]) == 4) {
    nblocks = N / nleaf; // 1 level
    admis = 1; // Strong admissibility
  }
  else if (atoi(argv[1]) == 5) {
    nblocks = 2; // Hierarchical (log_2(N/nleaf) levels)
    admis = 0; // Weak admissibility
  }
  else if (atoi(argv[1]) == 6) {
    nblocks = 2; // Hierarchical (log_2(N/nleaf) levels)
    admis = 1; // Strong admissibility
  }
  start("CPU compression");
  Hierarchical A(laplace1d, randx, N, N, rank, nleaf, admis, nblocks, nblocks);
  stop("CPU compression");
  rsvd_batch();
  //printXML(A);
  admis = N / nleaf; // Full rank
  start("Dense tree");
  //Hierarchical D(laplace1d, randx, N, N, rank, nleaf, admis, nblocks, nblocks);
  stop("Dense tree");
  Hierarchical x(random, randx, N, 1, rank, nleaf, admis, nblocks, 1);
  Hierarchical b(zeros, randx, N, 1, rank, nleaf, admis, nblocks, 1);
  //start("Verification");
  //double diff = (Dense(A) - Dense(D)).norm();
  //double norm = D.norm();
  //stop("Verification");
  //print("Compression Accuracy");
  //print("Rel. L2 Error", std::sqrt(diff/norm), false);
  print("Time");
  b.gemm(A, x, 1, 1);
  gemm_batch();
  stop("Init matrix");
  printTime("-DGEMM");
  start("LU decomposition");
  A.getrf();
  stop("LU decomposition");
  printTime("-DGETRF");
  printTime("-DTRSM");
  printTime("-DGEMM");
  start("Forward substitution");
  b.trsm(A,'l');
  stop("Forward substitution");
  printTime("-DTRSM");
  printTime("-DGEMM");
  start("Backward substitution");
  b.trsm(A,'u');
  stop("Backward substitution");
  printTime("-DTRSM");
  printTime("-DGEMM");
  double diff = (Dense(x) - Dense(b)).norm();
  double norm = x.norm();
  print("LU Accuracy");
  print("Rel. L2 Error", std::sqrt(diff/norm), false);
  return 0;
}

File: python/h_lu.py
import numpy as np
import sys

from FRANK.hierarchical import Hierarchical
from FRANK.dense import Dense
from FRANK.functions import (
    laplace1d,
    rand_data,
    zeros,
    Rand48
)
from FRANK.print import printf
from FRANK.timer import (
    start,
    stop
)


def main():
    np.set_printoptions(formatter={'float': '{: 0.15f}'.format}, linewidth=100)
    N = 128
    nleaf = 16
    rank = 4
    assert len(sys.argv) == 2
    nblocks = 0
    admis = 0
    randx = []
    rnd = Rand48(0)
    for i in range(N):
        randx.append(rnd.drand())
    randx = np.array(sorted(randx))
    start('Init Matrix')
    if (int(sys.argv[1]) == 0):
        # 1 level, full rank
        nblocks = int(N / nleaf)
        admis = int(N / nleaf)
    elif (int(sys.argv[1]) == 1):
        # log_2(N/nleaf) levels, full rank
        nblocks = 2
        admis = int(N / nleaf)
    elif (int(sys.argv[1]) == 2):
        # log_4(N/nleaf) levels, full rank
        nblocks = 4
        admis = int(N / nleaf)
    elif (int(sys.argv[1]) == 3):
        # 1 level, weak admissibility
        nblocks = int(N / nleaf)
        admis = 0
    elif (int(sys.argv[1]) == 4):
        # 1 level, strong admissibility
        nblocks = int(N / nleaf)
        admis = 1
    elif (int(sys.argv[1]) == 5):
        # log_2(N/nleaf) levels, weak admissibility
        nblocks = 2
        admis = 0
    elif (int(sys.argv[1]) == 6):
        # log_2(N/nleaf) levels, strong admissibility
        nblocks = 2
        admis = 1
    A = Hierarchical(randx, laplace1d, N, N, rank, nleaf, admis, nblocks, nblocks)
    admis = int(N / nleaf) # Full rank
    D = Hierarchical(randx, laplace1d, N, N, rank, nleaf, admis, nblocks, nblocks)
    x = Hierarchical(randx, rand_data, N, 1, rank, nleaf, admis, nblocks, 1)
    b = Hierarchical(randx, zeros, N, 1, rank, nleaf, admis, nblocks, 1)
    diff = (Dense(A) - Dense(D)).norm()
    norm = D.norm()
    printf('Compression Accuracy')
    printf('Rel. L2 Error', np.sqrt(diff/norm))
    printf('Time')
    b.gemm(A, x)
    stop('Init Matrix')
    start('LU decomposition')
    A.getrf()
    stop('LU decomposition')
    start('Forward substitution')
    b.trsm(A, 'l')
    stop('Forward substitution')
    start('Backward substitution')
    b.trsm(A, 'u')
    stop('Backward substitution')
    diff = (Dense(x) + Dense(b)).norm()
    norm = x.norm()
    printf('LU Accuracy')
    printf('Rel. L2 Error', np.sqrt(diff/norm))


if __name__ == '__main__':
    main()
